api/v1beta1/azuremachine_types.go

```diff
@@ -107,6 +107,11 @@ type AzureMachineSpec struct {
 	// +optional
 	AcceleratedNetworking *bool `json:"acceleratedNetworking,omitempty"`
 
+	// Diagnostics specifies the diagnostics settings for a virtual machine.
+	// If not specified the Boot diagnostics (Managed) will be enabled.
+	// +optional
+	Diagnostics Diagnostics `json:"diagnostics,omitempty"`
```

Comment:
Given this is optional, I think it needs to be a pointer else there's no chance for it to be omitted, and therefore will always require the additional fields within it that are required.

Have you been able to test the API behaviour to check that this doesn't cause required additional fields?

Perhaps because boot is a pointer it works that way. Perhaps the comment about if not specified should be on the `boot` level not this level

Reply:
Yes I manually tested this 👍
`Diagnostics` can be omitted with the current implementation, as you said, because of `Boot` being a pointer.

I can move `	// If not specified the Boot diagnostics (Managed) will be enabled.` down to the `BootDiagnostics` struct if we think it makes things clearer.

That said if people prefer a pointer here, I can change things up.

Reply:
How would be the behavior if `Diagnostics` field was defined as below? 
```suggestion
	Diagnostics *BootDiagnostics `json:"bootDiagnostics,omitempty"`
```

Would it be worth it to get rid of an extra struct of `Diagnostics` which internally is referring to `*Boot` ([Referring to Diagnostics struct here in `v1beta1/types.go`](https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2528/files#diff-964de9482de3e9bdd98a2e17c86bbc25650afa8b719a4c42a2a0014fc3b1869dR789-R796))

Reply:
How would be the behavior if `Diagnostics` field was defined as below? 
```suggestion
	Diagnostics *BootDiagnostics `json:"bootDiagnostics,omitempty"`
```

Would it be worth it to get rid of an extra struct of `Diagnostics` which internally is referring to `*Boot` ([Referring to Diagnostics struct here in `v1beta1/types.go`](https://github.com/kubernetes-sigs/cluster-api-provider-azure/pull/2528/files#diff-964de9482de3e9bdd98a2e17c86bbc25650afa8b719a4c42a2a0014fc3b1869dR789-R796))

Reply:
@nawazkh I don't think we would want `Diagnostics` to be limited to `BootDiagnostics` for two reasons:
- because in the future there might be more diagnostic areas, not only a Boot one, and so having an extra wrapping would serve as future proofing the API design
- the underlying `compute` API we are plugging into, follows this same pattern (probably for the same future proofing reason) `DiagnosticsProfile.BootDiagnostics`.

Reply:
Appreciate the thought about future proofing the API @damdo , Thanks! 

I was also slightly inclined on making `Diagnostics` a pointer as @JoelSpeed mentioned. 
However, we are able to _omit_ Diagnostics, if unspecified, which is a good(needed) behavior. 
I am ok with current implementation but would like to get @CecileRobertMichon 's opinion as well on this.

Reply:
> Diagnostics can be omitted with the current implementation, as you said, because of Boot being a pointer.

wouldn't that change if we add other fields inside Diagnostics in the future? Is there a good reason for _not_  making this a pointer? I think I'm +1 on pointer unless I'm missing something

Reply:
I'm happy to change it to a pointer for the reasons you three brought forward and for more consistency with similar fields.

Reply:
`Diagnostics` is now a pointer. PTAL